File: src/safety/safe-mode.ts
/**
 * Safe Mode - Safety controls for agent operations.
 *
 * This module implements safety guardrails to prevent the agent from making
 * unintended or dangerous changes. Safe mode is designed for production
 * environments where extra caution is required.
 *
 * ## Features
 *
 * 1. **Plan Requirement**: Mutating operations require a plan to be set first.
 *    This ensures the user has reviewed and approved the intended changes.
 *
 * 2. **Custom Validators**: Run shell commands after file changes to verify
 *    correctness (e.g., linters, type checkers, tests).
 *
 * 3. **LSP Diagnostic Blocking**: Block operations when LSP reports errors
 *    above a configurable severity threshold.
 *
 * ## Environment Variables
 *
 * - `MAESTRO_SAFE_MODE`: Set to "1" to enable safe mode
 * - `MAESTRO_SAFE_REQUIRE_PLAN`: Set to "0" to disable plan requirement
 * - `MAESTRO_SAFE_VALIDATORS`: Comma-separated list of validator commands
 * - `MAESTRO_SAFE_LSP_SEVERITY`: Max LSP severity to allow (1=error, 2=warning)
 *
 * ## Usage
 *
 * ```typescript
 * // Check if plan is required before mutating
 * requirePlanCheck("edit"); // Throws if safe mode enabled without plan
 *
 * // Run validators after successful file changes
 * await runValidatorsOnSuccess(["src/foo.ts"], lspDiagnostics);
 * ```
 */

import { exec } from "node:child_process";
import type { ExecException } from "node:child_process";
import { promisify } from "node:util";
import type { LspDiagnostic } from "../lsp/index.js";

const execAsync = promisify(exec);

// Environment variable names for configuration
const SAFE_MODE_ENV = "MAESTRO_SAFE_MODE";
const VALIDATORS_ENV = "MAESTRO_SAFE_VALIDATORS";
const REQUIRE_PLAN_ENV = "MAESTRO_SAFE_REQUIRE_PLAN";

/**
 * Internal state for safe mode configuration.
 * Configured once at startup and can be reconfigured for testing.
 */
type SafeModeState = {
	/** Whether safe mode is globally enabled */
	enabled: boolean;
	/** Whether a plan must be set before mutating operations */
	requirePlan: boolean;
	/** Shell commands to run as validators after file changes */
	validators: string[];
	/** LSP severity threshold (1=error only, 2=include warnings) */
	lspBlockingSeverity: number;
	/** Whether the plan requirement has been satisfied */
	planSatisfied: boolean;
};

/**
 * Result from running a validator command.
 * Captures output for reporting to the user.
 */
export type ValidatorRunResult = {
	/** The command that was executed */
	command: string;
	/** Standard output from the command */
	stdout: string;
	/** Standard error from the command */
	stderr: string;
};

/**
 * Error thrown when a validator fails.
 *
 * Contains the full validator output for debugging and reporting.
 * The agent should present this to the user and suggest fixes.
 */
export class ValidatorError extends Error {
	constructor(
		public readonly result: ValidatorRunResult,
		public readonly originalError?: Error,
	) {
		super(
			`Validator failed (${result.command}). See validator output for details.`,
		);
		this.name = "ValidatorError";
	}
}

/**
 * Module-level safe mode state.
 * Configured once at startup, can be reset for testing.
 */
const state: SafeModeState = {
	enabled: false,
	requirePlan: false,
	validators: [],
	lspBlockingSeverity: 1, // Default: only block on errors
	planSatisfied: false,
};

// Flag to prevent re-configuration after initial setup
let configured = false;

/**
 * Configure safe mode from environment variables.
 *
 * Called automatically at module load. Can be forced to reconfigure
 * for testing by passing force=true.
 *
 * @param force - If true, reconfigure even if already configured
 */
export function configureSafeMode(force = false): void {
	if (configured && !force) {
		return;
	}
	configured = true;

	// Check if safe mode is enabled via environment
	state.enabled = process.env[SAFE_MODE_ENV] === "1";
	state.requirePlan = false;
	state.validators = [];
	state.lspBlockingSeverity = process.env.MAESTRO_SAFE_LSP_SEVERITY
		? Number(process.env.MAESTRO_SAFE_LSP_SEVERITY)
		: 1;
	state.planSatisfied = false;

	// If not enabled, skip further configuration
	if (!state.enabled) {
		return;
	}

	// Plan requirement is enabled by default in safe mode unless explicitly disabled
	state.requirePlan = process.env[REQUIRE_PLAN_ENV] !== "0";

	// Parse comma-separated list of validator commands
	const validatorsRaw = process.env[VALIDATORS_ENV] ?? "";
	state.validators = validatorsRaw
		.split(",")
		.map((entry) => entry.trim())
		.filter((entry) => entry.length > 0);
}

// Auto-configure on module load
configureSafeMode();

/**
 * Mark the plan requirement as satisfied.
 *
 * Call this when the user has set up a plan (via todo list or plan mode).
 * This allows mutating operations to proceed.
 *
 * @param value - Whether the plan requirement is satisfied
 */
export function setPlanSatisfied(value: boolean): void {
	state.planSatisfied = value;
}

/**
 * Check if a plan is required before executing a tool.
 *
 * Throws an error if safe mode is enabled, plan is required,
 * and no plan has been set. Used by mutating tools (edit, write, bash).
 *
 * @param toolName - Name of the tool for the error message
 * @throws Error if plan check fails
 */
export function requirePlanCheck(toolName: string): void {
	// Skip check if safe mode disabled or plan not required
	if (!state.enabled || !state.requirePlan) {
		return;
	}
	// Skip check if plan has been set
	if (state.planSatisfied) {
		return;
	}
	throw new Error(
		`Safe mode requires a plan before executing ${toolName}. Create or update a todo checklist first.`,
	);
}

/** Extended error type from exec with captured output */
type ExecError = ExecException & { stdout?: string; stderr?: string };

/**
 * A diagnostic that blocks operation due to severity.
 */
type BlockingDiagnostic = {
	file: string;
	message: string;
	range?: LspDiagnostic["range"];
	severity?: number;
};

/**
 * Type guard to check if an error has stdout/stderr captured.
 */
function isExecError(error: unknown): error is ExecError {
	return (
		typeof error === "object" &&
		error !== null &&
		"stdout" in error &&
		"stderr" in error
	);
}

/**
 * Run validators after successful file changes.
 *
 * This function performs two types of validation:
 * 1. LSP diagnostics check - Fails if LSP reports errors above threshold
 * 2. Custom validators - Runs configured shell commands
 *
 * Validators receive the changed paths in MAESTRO_SAFE_CHANGED_PATHS env var
 * (separated by ::) so they can focus on relevant files.
 *
 * @param paths - Array of file paths that were changed
 * @param lspDiagnostics - Optional LSP diagnostics to check
 * @returns Array of successful validator results
 * @throws ValidatorError if any validation fails
 */
export async function runValidatorsOnSuccess(
	paths: string[],
	lspDiagnostics?: Record<string, LspDiagnostic[]>,
): Promise<ValidatorRunResult[]> {
	// Skip validation if safe mode is disabled
	if (!state.enabled) {
		return [];
	}

	const summaries: ValidatorRunResult[] = [];

	// Check LSP diagnostics first (before running expensive validators)
	if (lspDiagnostics) {
		const blocking = findBlockingDiagnostics(lspDiagnostics);
		if (blocking.length > 0) {
			// Format diagnostics as validator output
			const commandLabel = "lsp-diagnostics";
			throw new ValidatorError({
				command: commandLabel,
				stdout: blocking
					.map(
						(entry) =>
							`${entry.file}:${entry.range?.start.line ?? 0}:${entry.range?.start.character ?? 0} ${entry.message}`,
					)
					.join("\n"),
				stderr: "",
			});
		}
	}

	// Skip command validators if none configured
	if (state.validators.length === 0) {
		return summaries;
	}

	// Prepare environment with changed paths for validators
	const env = {
		...process.env,
		MAESTRO_SAFE_CHANGED_PATHS: paths.join("::"),
	};

	// Run each validator command sequentially
	for (const command of state.validators) {
		try {
			const { stdout, stderr } = await execAsync(command, {
				cwd: process.cwd(),
				env,
			});
			summaries.push({ command, stdout, stderr });
		} catch (error) {
			// Capture output from failed command
			if (isExecError(error)) {
				const summary: ValidatorRunResult = {
					command,
					stdout: error.stdout ?? "",
					stderr: error.stderr ?? "",
				};
				throw new ValidatorError(summary, error);
			}
			throw error;
		}
	}

	return summaries;
}

/**
 * Find LSP diagnostics that should block operation.
 *
 * Filters diagnostics by severity threshold configured in state.
 *
 * @param diagnostics - Map of file paths to their diagnostics
 * @returns Array of blocking diagnostics
 */
function findBlockingDiagnostics(
	diagnostics: Record<string, LspDiagnostic[]>,
): BlockingDiagnostic[] {
	const blocking: BlockingDiagnostic[] = [];
	for (const [file, entries] of Object.entries(diagnostics)) {
		for (const diag of entries) {
			// Treat missing severity as non-blocking (infinity)
			const severity = diag.severity ?? Number.POSITIVE_INFINITY;
			// Include if severity is at or below threshold (lower = more severe)
			if (severity <= state.lspBlockingSeverity) {
				blocking.push({
					file,
					message: diag.message,
					range: diag.range,
					severity: diag.severity,
				});
			}
		}
	}
	return blocking;
}

/**
 * Reset safe mode state for testing.
 *
 * Clears all configuration so tests can start fresh.
 */
export function resetSafeModeForTests(): void {
	configured = false;
	state.enabled = false;
	state.requirePlan = false;
	state.validators = [];
	state.planSatisfied = false;
}

/**
 * Check if safe mode is currently enabled.
 *
 * @returns true if safe mode is enabled
 */
export function isSafeModeEnabled(): boolean {
	return state.enabled;
}

File: packages/tui-rs/src/safety/safe_mode.rs
//! Safe mode guardrails (plan requirement + validators)
//!
//! Mirrors the TypeScript safe-mode gates at a minimal level:
//! - Require a plan before mutating operations (write/edit/bash/background tasks)
//! - Run configured validators after file mutations

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Mutex;

use crate::lsp::{self, LspDiagnostic};
use crate::tools::resolve_shell_config;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ValidatorResult {
    pub command: String,
    pub stdout: String,
    pub stderr: String,
    pub exit_code: i32,
}

#[derive(Debug, Clone)]
struct SafeModeConfig {
    enabled: bool,
    require_plan: bool,
    validators: Vec<String>,
    lsp_blocking_severity: u8,
}

static PLAN_SATISFIED: AtomicBool = AtomicBool::new(false);

static SAFE_MODE_CONFIG: std::sync::LazyLock<Mutex<SafeModeConfig>> =
    std::sync::LazyLock::new(|| {
        let enabled = std::env::var("MAESTRO_SAFE_MODE").ok().as_deref() == Some("1");
        let require_plan = if enabled {
            std::env::var("MAESTRO_SAFE_REQUIRE_PLAN").ok().as_deref() != Some("0")
        } else {
            false
        };
        let validators_raw = std::env::var("MAESTRO_SAFE_VALIDATORS").unwrap_or_default();
        let validators = validators_raw
            .split(',')
            .map(|entry| entry.trim().to_string())
            .filter(|entry| !entry.is_empty())
            .collect();
        let lsp_blocking_severity = std::env::var("MAESTRO_SAFE_LSP_SEVERITY")
            .ok()
            .and_then(|value| value.parse::<u8>().ok())
            .map_or_else(lsp::blocking_severity, |value| value.clamp(1, 4));

        Mutex::new(SafeModeConfig {
            enabled,
            require_plan,
            validators,
            lsp_blocking_severity,
        })
    });

/// Mark plan requirement as satisfied/unsatisfied.
pub fn set_plan_satisfied(value: bool) {
    PLAN_SATISFIED.store(value, Ordering::Relaxed);
}

/// Return true if safe mode is enabled.
pub fn is_safe_mode_enabled() -> bool {
    SAFE_MODE_CONFIG
        .lock()
        .map(|cfg| cfg.enabled)
        .unwrap_or(false)
}

/// Enforce plan requirement for mutating tools.
pub fn require_plan(tool_name: &str) -> Result<(), String> {
    let cfg = SAFE_MODE_CONFIG
        .lock()
        .map_err(|_| "Safe mode config unavailable".to_string())?;

    if !cfg.enabled || !cfg.require_plan {
        return Ok(());
    }

    if PLAN_SATISFIED.load(Ordering::Relaxed) {
        return Ok(());
    }

    Err(format!(
        "Safe mode requires a plan before executing {tool_name}. Create or update a todo checklist first."
    ))
}

/// Run validators configured via `MAESTRO_SAFE_VALIDATORS`.
pub async fn run_validators(paths: &[String]) -> Result<Vec<ValidatorResult>, String> {
    run_validators_with_diagnostics(paths, None).await
}

pub async fn run_validators_with_diagnostics(
    paths: &[String],
    lsp_diagnostics: Option<&HashMap<String, Vec<LspDiagnostic>>>,
) -> Result<Vec<ValidatorResult>, String> {
    let cfg = SAFE_MODE_CONFIG
        .lock()
        .map_err(|_| "Safe mode config unavailable".to_string())?
        .clone();

    if !cfg.enabled {
        return Ok(Vec::new());
    }

    if let Some(diagnostics) = lsp_diagnostics {
        let blocking = find_blocking_diagnostics(diagnostics, cfg.lsp_blocking_severity);
        if !blocking.is_empty() {
            let summary = blocking
                .into_iter()
                .map(|entry| {
                    format!(
                        "{}:{}:{} {}",
                        entry.file,
                        entry.line + 1,
                        entry.character + 1,
                        entry.message
                    )
                })
                .collect::<Vec<_>>()
                .join("\n");
            return Err(format!(
                "LSP diagnostics blocked safe-mode validators:\n{summary}"
            ));
        }
    }

    if cfg.validators.is_empty() {
        return Ok(Vec::new());
    }

    let (shell, shell_args) = resolve_shell_config()
        .map_err(|e| format!("Failed to resolve shell for validators: {e}"))?;

    let mut results = Vec::new();
    for command in cfg.validators {
        let mut cmd = tokio::process::Command::new(&shell);
        cmd.args(&shell_args)
            .arg(command.clone())
            .env("MAESTRO_SAFE_CHANGED_PATHS", paths.join("::"));

        let output = cmd
            .output()
            .await
            .map_err(|e| format!("Failed to run validator '{command}': {e}"))?;

        let stdout = String::from_utf8_lossy(&output.stdout).to_string();
        let stderr = String::from_utf8_lossy(&output.stderr).to_string();
        let exit_code = output.status.code().unwrap_or(-1);

        let result = ValidatorResult {
            command: command.clone(),
            stdout,
            stderr,
            exit_code,
        };

        if !output.status.success() {
            return Err(format!(
                "Validator failed ({}). Exit code: {}. Stderr: {}",
                command,
                exit_code,
                result.stderr.trim()
            ));
        }

        results.push(result);
    }

    Ok(results)
}

#[derive(Debug, Clone)]
struct BlockingDiagnostic {
    file: String,
    line: u32,
    character: u32,
    message: String,
}

fn find_blocking_diagnostics(
    diagnostics: &HashMap<String, Vec<LspDiagnostic>>,
    threshold: u8,
) -> Vec<BlockingDiagnostic> {
    let mut blocking = Vec::new();
    for (file, entries) in diagnostics {
        for diag in entries {
            let severity = diag.severity.unwrap_or(u8::MAX);
            if severity <= threshold {
                blocking.push(BlockingDiagnostic {
                    file: file.clone(),
                    line: diag.range.start.line,
                    character: diag.range.start.character,
                    message: diag.message.clone(),
                });
            }
        }
    }
    blocking
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lsp::{LspPosition, LspRange};

    // ========================================================================
    // ValidatorResult Tests
    // ========================================================================

    #[test]
    fn test_validator_result_serialization() {
        let result = ValidatorResult {
            command: "npm test".to_string(),
            stdout: "All tests passed".to_string(),
            stderr: String::new(),
            exit_code: 0,
        };
        let json = serde_json::to_value(&result).unwrap();
        assert_eq!(json["command"], "npm test");
        assert_eq!(json["exit_code"], 0);
    }

    #[test]
    fn test_validator_result_with_error() {
        let result = ValidatorResult {
            command: "cargo clippy".to_string(),
            stdout: String::new(),
            stderr: "error: unused variable".to_string(),
            exit_code: 1,
        };
        assert_eq!(result.exit_code, 1);
        assert!(result.stderr.contains("unused variable"));
    }

    // ========================================================================
    // Plan Satisfied Tests
    // ========================================================================

    #[test]
    fn test_set_plan_satisfied() {
        // Reset to known state
        set_plan_satisfied(false);
        assert!(!PLAN_SATISFIED.load(std::sync::atomic::Ordering::Relaxed));

        set_plan_satisfied(true);
        assert!(PLAN_SATISFIED.load(std::sync::atomic::Ordering::Relaxed));

        set_plan_satisfied(false);
        assert!(!PLAN_SATISFIED.load(std::sync::atomic::Ordering::Relaxed));
    }

    // ========================================================================
    // BlockingDiagnostic Tests
    // ========================================================================

    #[test]
    fn test_find_blocking_diagnostics_empty() {
        let diagnostics: HashMap<String, Vec<LspDiagnostic>> = HashMap::new();
        let blocking = find_blocking_diagnostics(&diagnostics, 2);
        assert!(blocking.is_empty());
    }

    #[test]
    fn test_find_blocking_diagnostics_with_error() {
        let mut diagnostics = HashMap::new();
        diagnostics.insert(
            "src/main.rs".to_string(),
            vec![LspDiagnostic {
                range: LspRange {
                    start: LspPosition {
                        line: 10,
                        character: 5,
                    },
                    end: LspPosition {
                        line: 10,
                        character: 15,
                    },
                },
                message: "undefined variable".to_string(),
                severity: Some(1), // Error
                source: None,
            }],
        );

        // Threshold 2 (warning) should include errors (severity 1)
        let blocking = find_blocking_diagnostics(&diagnostics, 2);
        assert_eq!(blocking.len(), 1);
        assert_eq!(blocking[0].file, "src/main.rs");
        assert_eq!(blocking[0].line, 10);
        assert_eq!(blocking[0].character, 5);
        assert_eq!(blocking[0].message, "undefined variable");
    }

    #[test]
    fn test_find_blocking_diagnostics_filters_by_severity() {
        let mut diagnostics = HashMap::new();
        diagnostics.insert(
            "src/lib.rs".to_string(),
            vec![
                LspDiagnostic {
                    range: LspRange {
                        start: LspPosition {
                            line: 1,
                            character: 0,
                        },
                        end: LspPosition {
                            line: 1,
                            character: 10,
                        },
                    },
                    message: "error message".to_string(),
                    severity: Some(1), // Error
                    source: None,
                },
                LspDiagnostic {
                    range: LspRange {
                        start: LspPosition {
                            line: 5,
                            character: 0,
                        },
                        end: LspPosition {
                            line: 5,
                            character: 20,
                        },
                    },
                    message: "hint message".to_string(),
                    severity: Some(4), // Hint
                    source: None,
                },
            ],
        );

        // Threshold 1 (error only) should only include the error
        let blocking = find_blocking_diagnostics(&diagnostics, 1);
        assert_eq!(blocking.len(), 1);
        assert_eq!(blocking[0].message, "error message");

        // Threshold 4 should include both
        let all = find_blocking_diagnostics(&diagnostics, 4);
        assert_eq!(all.len(), 2);
    }

    #[test]
    fn test_find_blocking_diagnostics_default_severity() {
        let mut diagnostics = HashMap::new();
        diagnostics.insert(
            "src/test.rs".to_string(),
            vec![LspDiagnostic {
                range: LspRange {
                    start: LspPosition {
                        line: 0,
                        character: 0,
                    },
                    end: LspPosition {
                        line: 0,
                        character: 0,
                    },
                },
                message: "no severity".to_string(),
                severity: None, // Will default to u8::MAX
                source: None,
            }],
        );

        // With any normal threshold, diagnostics without severity should not block
        let blocking = find_blocking_diagnostics(&diagnostics, 4);
        assert!(blocking.is_empty());
    }

    #[test]
    fn test_find_blocking_diagnostics_multiple_files() {
        let mut diagnostics = HashMap::new();
        diagnostics.insert(
            "src/a.rs".to_string(),
            vec![LspDiagnostic {
                range: LspRange {
                    start: LspPosition {
                        line: 1,
                        character: 0,
                    },
                    end: LspPosition {
                        line: 1,
                        character: 5,
                    },
                },
                message: "error in a".to_string(),
                severity: Some(1),
                source: None,
            }],
        );
        diagnostics.insert(
            "src/b.rs".to_string(),
            vec![LspDiagnostic {
                range: LspRange {
                    start: LspPosition {
                        line: 2,
                        character: 0,
                    },
                    end: LspPosition {
                        line: 2,
                        character: 5,
                    },
                },
                message: "error in b".to_string(),
                severity: Some(1),
                source: None,
            }],
        );

        let blocking = find_blocking_diagnostics(&diagnostics, 2);
        assert_eq!(blocking.len(), 2);
    }
}
